### ISSUE TITLE
How does InplaceInferer works?

### ISSUE THREAD

**Issue by jczaja** [PaddlePaddle/Paddle#23074]
@luotao1 , @sneaxiy 

Hi We are interested in taking adventage of inplace optimizations in DNNL integration (starting with softmax). PaddlePaddle inplace mechanism is for example used here:
https://github.com/PaddlePaddle/Paddle/blob/01ab8a061936aee264c8e9aeae360cf3de195e47/paddle/fluid/operators/softmax_op.cc#L231-L235

1. How does it work?
2. Is softmax to work inplace always ?
3. How can I detect inside kernel that kernel is to work inplace?

looking forward to your answers

**Comment by sneaxiy** [comment]
Hi, @jczaja 

**1. How does it work?**
The in-place strategy would be enabled by default when you use `ParallelExecutor` or use `Executor` to run `CompiledProgram`. It is not supported when you use `Executor` to run `Program`. Inside `ParallelExecutor`, we use a pass called `BufferSharedInplaceOpPass` to detect whether in-place can be performed. 

For example,  x_var -> op -> y_var, **only if (1) x_var is not used in the following network, (2) and the calculation result would be always right when y_var and x_var shares the same memory, we would perform in-place before running this op**.

However, the framework can only tell us whether x_var is not used in the following network, and there is no way to know whether the calculation result would be alway right when y_var and x_var share the same memory. `ops::SoftmaxInplaceInferer` is a helper class for each op, which would tell the framework which in/out variable pairs can be in-placed without affecting the calculation results. In this way, the framework can know which two variables can be in-placed.

The way we perform in-place for in/out is to call `Tensor::ShareBufferWith` method, which would share the underlying memory space of 2 tensors. Notice that y_var and x_var would always be 2 different variables no matter whether in-place is performed or not. They would only share memory space when in-place is performed. That is to say, x_var and y_var may have different names, shapes, lods, etc.

**2. Is softmax to work inplace always?**
The answer is no. As mentioned above, whether in-place can be performed is detected automatically by the framework itself.

**3. How can I detect inside kernel that kernel is to work inplace?**
As mentioned before, the way we perform in-place is to call `Tensor::ShareBufferWith` methods. If you want to know whether two tensors are in-placed, you can compare their `Tensor::Holder()`. If their holders are the same, in-place is exactly performed.

**Comment by jczaja** [comment]
Hi @sneaxiy ,

1. My goal is to enable C-API (Predictor , NaiveExecutor) to have softmax inplace execution 
I have tried to use mentioned pass:
https://github.com/PaddlePaddle/Paddle/blob/31fc3ab79a75631b26883585b1bb25af0134a0eb/paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_inplace_op_pass.cc#L156-L160
by adding it to the other passes MKL-DNN passes declared here:
https://github.com/PaddlePaddle/Paddle/blob/abee05a8c877bcb24bf15ebf582690bd85558d6b/paddle/fluid/inference/api/paddle_pass_builder.cc#L186-L200

But when running test_analyzer_bert I got an error:
`Error: Pass buffer_shared_inplace_pass has not been registered at (/home/jczaja/paddle/paddle/fluid/framework/ir/pass.h:185)
`
I'm running C-API , prediction using MKL-DNN , no GPU builds. Could you please advice how can I have inplace softmax enabled for this scenario(CPU, C-API, mkl-dnn)?

2. Is it required that all kernels of specific type should support inplace execution ? For example
sum op CPU is having support for inplace, but sum MKL-DNN kernel op does not support inplace.
Is it possible not to have all sum kernels to support inplace Or if inplace is enabled then all kernels should be enable to do inplace computation?

**Comment by sneaxiy** [comment]
Hi, @jczaja 

1. `BufferSharedInplaceOpPass` is not suitable for C-API prediction, it is designed for `ParallelExecutor`. If you want to perform in-place in C-API, you may need to add your own in-place implementation including lifecycle analysis of each var and perform `Tensor::ShareBufferWith` before running each op.

2. In the current implementation, all kernels of specific type should support inplace execution. However, we can change this limitation, and make some kernels support inplace while the others does not. Does MKL-DNN sum kernel not support in-place? If so, I would fix this bug asap.

**Comment by jczaja** [comment]
Hi @sneaxiy,
1. Ok, I will try to implement our own inplace pass
2. Currently MKL-DNN kernels are **not** set to inplace. But some MKL-DNN primitives  support inplace:
- eltwise (activations)
- binary (elementwise_add, elementwise_mul)
- layer_norm (to be done)
- batch_norm
- softmax

sum of MKL-DNN is not supporting inplace, but for the purpose of sum op of Paddle we can use
binary primitve of MKL-DNN which should do the job. so that one would require a bit more modifications.

So we do not have those inplace optimizations enabled.
By adding our own inplace pass as You suggessted we can control for which operators
inplace is to be enabled , and later when adapt MKL-DNN kernels we can add new types to inplace list.
but this is  for C-API

3. For parallel executor, We should probably enable inplace of MKL-DNN kernels as soon as possible.
   Is in-place enabled in parallel executor  for inference only or just for both inference and training? 

4. Is there any model or unit test running parallel executor where I can test inplace pass ?

**Comment by sneaxiy** [comment]
Hi, @jczaja .

> sum of MKL-DNN is not supporting inplace, but for the purpose of sum op of Paddle we can use binary primitve of MKL-DNN which should do the job. so that one would require a bit more modifications.

I found that in sum mkldnn kernel, we judge whether in-place is happened by comparing whether `out_var` is the same var as `in_vars[0]`. How about change it to be `out_var->Get<framework::LoDTensor>().Holder() == in_vars[0]->Get<framework::LoDTensor>().Holder()` , so that we can support in-place in sum mkldnn kernel?

https://github.com/PaddlePaddle/Paddle/blob/3af477112281a06bd6824945ca9e7c78015cbd66/paddle/fluid/operators/mkldnn/sum_mkldnn_op.cc#L61


> For parallel executor, We should probably enable inplace of MKL-DNN kernels as soon as possible. Is in-place enabled in parallel executor for inference only or just for both inference and training?

Yes, in-place strategy is enabled by default in ParallelExecutor, no matter the network is for training or inference.

> Is there any model or unit test running parallel executor where I can test inplace pass ?

Since parallel executor enables in-place by default, you can use any codes to test the in-place pass. For example, if you want to test relu, you can write the following network:

```
x = fluid.data(shape=xxx, ...)
y = fluid.layers.tanh(x)
z = fluid.layers.relu(y)
```
**Notice that we do not perform in-place for the feeded variables.** Because the users may use the feeded variables for other purpose, such as feed it into another network, the feeded variables should not be changed. So in the example above, in-place would not happen in x->tanh->y even though x does not occur in the following network, but happen in y->relu->z.

**Comment by jczaja** [comment]
Hi @sneaxiy ,

Currently I'm working on C-API mkl-dnn in-place pass . So I have questions:
1) Is it fine if I use renaming for in-place C-APO pass implementation? You advise against it :
https://github.com/PaddlePaddle/Paddle/blob/430b0099c9e6015514a4c5bb7f7fc6845fe4731b/paddle/fluid/framework/ir/memory_optimize_pass/memory_reuse_pass.h#L34-L72
But C-API is a simplified API e.g.  only inference, single-node (no distributed communication), user is mostly executing program based on loaded model . So is it fine that We implement mkl-dnn inplace pass using renaming ? 
~~2) How can I get InplaceInferer object in C-API pass?~~

**Comment by jczaja** [comment]
Thanks for all answers and explanation
